elpy-django.el

```diff
@@ -168,27 +168,43 @@ test arguments in `elpy-django-test-runner-args'."
     (setq dj-commands-str (mapcar (lambda (x) (s-trim x)) dj-commands-str))
     (sort dj-commands-str 'string-lessp)))
 
+
+(defvar elpy-django--test-runner-cache '()
+  "Internal cache for elpy-django--get-test-runner.
+The cache is keyed on project root and DJANGO_SETTINGS_MODULE env var")
+
+
 (defun elpy-django--get-test-runner ()
   "Return the name of the django test runner.
-Needs `DJANGO_SETTINGS_MODULE' to be set in order to work."
+Needs `DJANGO_SETTINGS_MODULE' to be set in order to work.
+The result is memoized on project root and `DJANGO_SETTINGS_MODULE'"
   (let ((django-import-cmd "import django;django.setup();from django.conf import settings;print(settings.TEST_RUNNER)")
         (django-settings-env (getenv "DJANGO_SETTINGS_MODULE"))
         (default-directory (elpy-project-root)))
     ;; If no Django settings has been set, then nothing will work. Warn user
     (when (not django-settings-env)
       (error "Please set environment variable `DJANGO_SETTINGS_MODULE' if you'd like to run the test runner"))
 
-    ;; We have to be able to import the DJANGO_SETTINGS_MODULE to detect test
-    ;; runner; if python process importing settings exits with error,
-    ;; then warn the user that settings is not valid
-    (unless (= 0 (call-process elpy-rpc-python-command nil nil nil
-                               "-c" (format "import %s" django-settings-env)))
-      (error (format "Unable to import DJANGO_SETTINGS_MODULE: '%s'"
-                     django-settings-env)))
-
-    ;; Return test runner
-    (s-trim (shell-command-to-string
-             (format "%s -c '%s'" elpy-rpc-python-command django-import-cmd)))))
+    (let* ((runner-key (list default-directory django-settings-env))
```

Comment:
Can you extract two functions here to make this look like the following? Makes the code more readable I think.

```Lisp
(or (elpy-django--get-test-runner-from-cache)
    (elpy-django--cache-test-runner
     (elpy-django--determine-test-runner)))
```

Reply:
Done, this is indeed more readable.